File: 2024/day05_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Rule = tuple[int, int]


def solution(content: str) -> Any:
    [fst, snd] = content.split("\n\n")
    rules = parse_rules(fst)
    pages = parse_pages(snd)
    return sum(
        middle(correct(page_group, rules))
        for page_group in pages
        if any(is_invalid(rule, page_group) for rule in rules)
    )


def correct(pages: list[int], rules: list[Rule]) -> list[int]:
    for rule in rules:
        if is_invalid(rule, pages):
            [fst, snd] = rule
            corrected = pages.copy()
            fst_index = corrected.index(fst)
            snd_index = corrected.index(snd)
            corrected[fst_index], corrected[snd_index] = (
                corrected[snd_index],
                corrected[fst_index],
            )
            return correct(corrected, rules)
    return pages


def middle(pages: list[int]) -> int:
    if not pages:
        return 0
    return pages[len(pages) // 2]


def is_invalid(rule: Rule, pages: list[int]) -> bool:
    [fst, snd] = rule
    if fst not in pages or snd not in pages:
        return False
    return pages.index(fst) > pages.index(snd)


def parse_rules(content: str) -> list[Rule]:
    return [
        (int(fst), int(snd))
        for [fst, snd] in [line.split("|") for line in content.split("\n")]
    ]


def parse_pages(content: str) -> list[list[int]]:
    return [[int(page) for page in line.split(",")] for line in content.split("\n")]


def main() -> None:
    with open("day05.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()

File: 2024/day05-2.ts
type Rule = {
  before: number;
  after: number;
};

function solution(content: string) {
  const [fst, snd] = content.split("\n\n");
  const rules = parseRules(fst);
  const pages = parsePages(snd);
  return pages
    .filter((group) => anyInvalid(rules, group))
    .map((group) => correct(group, rules))
    .map(middle)
    .reduce((acc, cur) => acc + cur, 0);
}

function correct(pages: number[], rules: Rule[]) {
  const copy = pages.slice();
  while (true) {
    const wrong = rules.find((rule) => isInvalid(rule, copy));
    if (wrong) {
      const { before, after } = wrong;
      const beforeIndex = copy.indexOf(before);
      const afterIndex = copy.indexOf(after);
      const temp = copy[beforeIndex];
      copy[beforeIndex] = copy[afterIndex];
      copy[afterIndex] = temp;
    } else {
      break;
    }
  }
  return copy;
}

function middle(pages: number[]) {
  return pages.at(pages.length / 2)!;
}

function anyInvalid(rules: Rule[], pages: number[]) {
  return rules.some((rule) => isInvalid(rule, pages));
}

function isInvalid(rule: Rule, pages: number[]) {
  const { before, after } = rule;
  if (!pages.includes(before) || !pages.includes(after)) return false;
  return pages.indexOf(before) > pages.indexOf(after);
}

function parseRules(content: string): Rule[] {
  return content.split("\n").map((line) => {
    const [before, after] = line.split("|").map(Number);
    return { before, after };
  });
}

function parsePages(content: string): number[][] {
  return content.split("\n").map((line) => line.split(",").map(Number));
}

Deno.readTextFile("day05.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
